zeppelin-web/src/app/jobmanager/jobmanager.component.js

```diff
@@ -0,0 +1,181 @@
+    const cp = $scope.pagination.currentPage
+    const itp = $scope.pagination.itemsPerPage
+    return jobs.slice((cp - 1) * itp, (cp * itp))
+  }
+
+  let asyncNotebookJobFilter = function (jobs, filterConfig) {
+    return new Promise((resolve, reject) => {
+      $scope.filteredJobs = jobManagerFilter(jobs, filterConfig)
+      resolve($scope.filteredJobs)
+    })
+  }
+
+  $scope.$watch('sorter.currentDateSorter', function() {
+    $scope.filterConfig.isSortByAsc =
+      $scope.sorter.currentDateSorter === JobDateSorter.OLDEST_UPDATED
+    asyncNotebookJobFilter($scope.jobs, $scope.filterConfig)
+  })
+
+  $scope.getJobIconByStatus = getJobIconByStatus
+  $scope.getJobColorByStatus = getJobColorByStatus
+
+  $scope.filterJobs = function (jobs, filterConfig) {
+    asyncNotebookJobFilter(jobs, filterConfig)
+      .then(() => {
+        $scope.isFilterLoaded = true
+      })
+      .catch(error => {
+        console.error('Failed to search jobs from server', error)
+      })
+  }
+
+  $scope.filterValueToName = function (filterValue, maxStringLength) {
+    if (typeof $scope.defaultInterpreters === 'undefined') {
+      return
+    }
+
+    let index = $scope.defaultInterpreters.findIndex(intp => intp.value === filterValue)
+    if (typeof $scope.defaultInterpreters[index].name !== 'undefined') {
+      if (typeof maxStringLength !== 'undefined' &&
+        maxStringLength > $scope.defaultInterpreters[index].name) {
+        return $scope.defaultInterpreters[index].name.substr(0, maxStringLength - 3) + '...'
+      }
+      return $scope.defaultInterpreters[index].name
+    } else {
+      return 'NONE'
+    }
+  }
+
+  $scope.setFilterValue = function (filterValue) {
+    $scope.filterConfig.interpreterFilterValue = filterValue
+    $scope.filterJobs($scope.jobs, $scope.filterConfig)
+  }
+
+  $scope.setJobs = function(jobs) {
+    $scope.jobs = jobs
+    let interpreters = $scope.jobs
+      .filter(j => typeof j.interpreter !== 'undefined')
+      .map(j => j.interpreter)
+    interpreters = [...new Set(interpreters)] // remove duplicated interpreters
+
+    $scope.defaultInterpreters = [ { name: 'ALL', value: '*' } ]
+    for (let i = 0; i < interpreters.length; i++) {
+      $scope.defaultInterpreters.push({ name: interpreters[i], value: interpreters[i] })
+    }
+  }
+
+  function init() {
+    websocketMsgSrv.getNoteJobsList()
+
+    $scope.$on('$destroy', function () {
+      websocketMsgSrv.unsubscribeJobManager()
+    })
+  }
+
+  /*
+   ** $scope.$on functions below
+   */
+
+  $scope.$on('setNoteJobs', function (event, responseData) {
+    $scope.setJobs(responseData.jobs)
+    $scope.filterJobs($scope.jobs, $scope.filterConfig)
+  })
+
+  $scope.$on('setUpdateNoteJobs', function (event, responseData) {
+    let jobs = $scope.jobs
+    let jobByNoteId = jobs.reduce((acc, j) => {
+      const noteId = j.noteId
+      acc[noteId] = j
+      return acc
+    }, {})
+
+    let notes = responseData.jobs
```

Comment:
I scrolled down and saw this is from `jobmanager.controller.js`. @1ambda, what do you think? Is it a good idea to change the code here?

IMO, The logic here seems a bit complicated. A lot of side effects are happening inside `notes.map`
It would be helpful, if we can simplify the logic or have a small one line comment describing what's happening